Objects/weakrefobject.c

```diff
@@ -48,6 +128,59 @@ new_weakref(PyObject *ob, PyObject *callback)
     return result;
 }
 
+#ifdef Py_GIL_DISABLED
+
+// Clear the weakref and steal its callback into `callback`, if provided.
+static void
+clear_weakref_lock_held(PyWeakReference *self, PyObject **callback)
+{
+    // TODO: Assert locks are held or world is stopped
+    if (_Py_atomic_load_ptr(&self->wr_object) != Py_None) {
+        PyWeakReference **list = GET_WEAKREFS_LISTPTR(self->wr_object);
+        if (*list == self)
+            /* If 'self' is the end of the list (and thus self->wr_next ==
+               NULL) then the weakref list itself (and thus the value of *list)
+               will end up being set to NULL. */
+            _Py_atomic_store_ptr(list, self->wr_next);
```

Comment:
We read `wr_next` without holding the lock at the beginning of `PyObject_ClearWeakRefs` to provide a fast path for the common case where an object doesn't have any weakrefs. Since `PyObject_ClearWeakRefs` is called as part of the deallocation path for any object that *could* have weakrefs I think it makes sense to have a fast path.

Reply:
Ok, that makes sense to me.